### PR TITLE
Add support for paramaterizing varargs in PyCharm

### PR DESCRIPTION
**Pull request by kroo**
Previously, `*args` and `**kwargs` were skipped when matching against parameterized types in the type checker.  This prevented pycharm from understanding type hints on functions whose return type are a function of the varardic parameters.

This can be reproduced with the following code:

``` python
def f(*args):
    """
    :type args: tuple[T]
    :rtype: T
    """
    return args[0]

expr = f(1)  # expect expr to be of type "int"
```

### CODE REVIEW

**Comment by klikh** [comment]
Have you submitted a contributor license agreement? If not, please download the PDF from http://www.jetbrains.org/display/IJOS/Contributor+Agreement and then print it out, sign (mention you github account name there), scan and email to contribute@jetbrains.com

**Comment by kroo** [comment]
Sent!

**Comment by kroo** [comment]
@klikh, is there anything I can do to help this PR along?

**Comment by klikh** [comment]
I've forwarded your request to the PyCharm developer. He'll look at this one soon. Sorry for delay.

**Comment by kroo** [comment]
@klikh Anything I should do to bump this again?

**Comment by vlasovskikh** [comment]
@kroo Thanks for your PR! Sorry for the late reply. 

We would like the type hints notation for varargs to be compatible with PEP 0484, where you don't have to state that `*args` is a tuple of some type; you just specify the type of the elements in that tuple. For example, we would like this:

```python
def f(*args, **kwargs):
    """
    :type args: T
    :type kwargs: V
    :rtype: (T, V)
    """
    return args[0], kwargs['foo']
```

to be similar to PEP 0484 syntax:

```python
T = typing.TypeVar('T')
V = typing.TypeVar('V')

def f(*args: T, **kwargs: V):
    return args[0], kwargs['foo']
```

File: python/src/com/jetbrains/python/psi/types/PyTypeChecker.java [hunk 1]
```diff
@@ -393,11 +393,26 @@ else if (type instanceof PyCallableType) {
...
       final PyType paramType = context.getType(p);
+
+      PyType argType = context.getType(entry.getKey());
+      if (p.isPositionalContainer()) {
+        final PyClassType tupleClassType = PyBuiltinCache.getInstance(p).getTupleType();
+        if (tupleClassType == null) {
+          return null;
+        }
+        final PyClass tupleClass = tupleClassType.getPyClass();
+        argType = new PyCollectionTypeImpl(tupleClass, false, argType);
```
Comment (vlasovskikh) [inline python/src/com/jetbrains/python/psi/types/PyTypeChecker.java]: We now have `new PyTupleType(tupleClass, elementTypes, true)` for homogeneous tuples in our type system.

**Comment by kroo** [comment]
Thanks, this makes sense.  I'll give it a go.

**Comment by sproshev** [comment]
Fixes related to PY-19723 are in master.

**Pull request closed without merging**